server/controllers/admin/roles.js

```diff
@@ -0,0 +1,223 @@
+const db = require('../../lib/db');
+const q = require('q');
+const _ = require('lodash');
+
+module.exports.list = list;
+module.exports.detail = detail;
+module.exports.create = create;
+module.exports.update = update;
+module.exports.remove = remove;
+module.exports.affectPages = affectPages;
+module.exports.affectToUser = affectToUser;
+module.exports.listForUser = listForUser;
+
+function list(req, res, next) {
+  const sql = `
+    SELECT BUID(uuid) as uuid, label, project_id 
+    FROM role
+    WHERE project_id=?
+  `;
+  db.exec(sql, req.params.project_id)
+    .then((rows) => {
+      res.json(rows);
+    }).catch(next)
+    .done();
+}
+
+
+
+function detail(req, res, next) {
+
+  const sql = `
+   SELECT BUID(uuid) as uuid, label, project_id 
+   FROM role
+   WHERE  uuid = HUID(?) LIMIT 1
+  `;
+
+  db.one(sql, req.params.uuid)
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+// create a new role
+function create(req, res, next) {
+
+  var sql = `
+    INSERT INTO  role(uuid, label, project_id)
+    VALUES( HUID(uuid()), ?,?)
+  `;
+
+  db.exec(sql, [req.body.label, req.body.project_id])
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+function update(req, res, next) {
+  const role = req.body;
+  delete role.uuid;
+  const sql = ` UPDATE role SET ? WHERE uuid = HUID(?)`;
+
+  db.exec(sql, [role, req.params.uuid])
+    .then((rows) => {
+      res.json(rows);
+    })
+  .catch(next)
+    .done();
+}
+
+
+function remove(req, res, next) {
+  const sql = `DELETE FROM role WHERE uuid = HUID(?)`;
+  db.exec(sql, req.params.uuid)
+    .then((rows) => {
+      res.json(rows);
+    })
+    .catch(next)
+    .done();
+}
+
+// affect permission to a specific role
+
+function affectPages(req, res, next) {
+  const data = req.body;
+
+  const unitIds = data.unit_ids;
+  const roleUuid = data.role_uuid;
+  var msg = {
+    state : 1,
+    message : 'success',
+  };
+  const deleteFromRole = `DELETE FROM role_unit WHERE role_uuid = HUID(?) `;
+  const affectPage = `
+    INSERT INTO  role_unit(uuid, unit_id, role_uuid)
+    VALUES( HUID(uuid()), ?, HUID(?))`;
```

Comment:
The SQL `uuid()` function is a version 1 uuid, while the uuid generated by our JS functions are version 4 uuids.  For broad overview of the differences, see [this SO answer](https://stackoverflow.com/questions/20342058/which-uuid-version-to-use/20342413#20342413).  It's best that we keep using the same version 4 uuid.

For a more general overview, [Wikipedia](https://en.wikipedia.org/wiki/Universally_unique_identifier) has a good summary.